proof.ml

```diff
@@ -517,3 +521,209 @@ let rec get_proofs_of_sequents proof =
+    | Par_proof (head, e1, e2, tail, Weakening_proof (head_wk1, f1, tail_wk1, Weakening_proof (head_wk2, f2, tail_wk2, p)))
+        when List.length head_wk2 <> List.length head ->
+        (* If first weaken after par is on e1 or e2 and if second weaken is not on the other one,
+            then we can commute second weaken with {par + first weaken}. *)
+        let new_head_wk2, new_tail_wk2, new_head_wk1, new_tail_wk1 =
+            new_head_wk_tail_wk_head_tail head_wk1 tail_wk1 head_wk2 tail_wk2 (Whynot f1) 0 in
+        let second_weakening = Weakening_proof (new_head_wk1, f1, new_tail_wk1, p) in
+        let new_head_wk2, new_tail_wk2, new_head, new_tail =
+            new_head_wk_tail_wk_head_tail head tail new_head_wk2 new_tail_wk2 (Par (e1, e2)) 2 in
+        let new_proof = get_commuted_proof (rec_commute_down_weakenings (Par_proof (new_head, e1, e2, new_tail, second_weakening))) in
+        true, Weakening_proof (new_head_wk2, f2, new_tail_wk2, new_proof)
+    | Par_proof (head, e1, e2, tail, Weakening_proof (head_wk1, f1, tail_wk1, Weakening_proof (head_wk2, f2, tail_wk2, Weakening_proof (head_wk3, f3, tail_wk3, p)))) ->
+        (* If first weaken after par is on e1 or e2 and second weaken is on the other one (cf cases before).
+            We can commute third weaken with {par + first weaken + second weaken}. *)
+        let new_head_wk3, new_tail_wk3, new_head_wk2, new_tail_wk2 =
+            new_head_wk_tail_wk_head_tail head_wk2 tail_wk2 head_wk3 tail_wk3 (Whynot f2) 0 in
+        let third_weakening = Weakening_proof (new_head_wk2, f2, new_tail_wk2, p) in
+        let new_head_wk3, new_tail_wk3, new_head_wk1, new_tail_wk1 =
+            new_head_wk_tail_wk_head_tail head_wk1 tail_wk1 new_head_wk3 new_tail_wk3 (Whynot f1) 0 in
+        let second_weakening = Weakening_proof (new_head_wk1, f1, new_tail_wk1, third_weakening) in
+        let new_head_wk3, new_tail_wk3, new_head, new_tail =
+            new_head_wk_tail_wk_head_tail head tail new_head_wk3 new_tail_wk3 (Par (e1, e2)) 2 in
+        let new_proof = get_commuted_proof (rec_commute_down_weakenings (Par_proof (new_head, e1, e2, new_tail, second_weakening))) in
+        true, Weakening_proof (new_head_wk3, f3, new_tail_wk3, new_proof)
+    | With_proof (head, e1, e2, tail, Weakening_proof (head_wk1, f1, tail_wk1, p1), Weakening_proof (head_wk2, f2, tail_wk2, p2))
+        when List.length head_wk1 = List.length head_wk2 && List.length head_wk1 <> List.length head ->
+        (* f1 and f2 should be equal, head_wk1 and head_wk2, tail_wk1 and tail_wk2 should be equal if you replace e1 by e2 *)
+        let new_head_wk, new_tail_wk, new_head, new_tail =
+            new_head_wk_tail_wk_head_tail head tail head_wk1 tail_wk1 (With (e1, e2)) 1 in
+        let new_proof = get_commuted_proof (rec_commute_down_weakenings (With_proof (new_head, e1, e2, new_tail, p1, p2))) in
+        true, Weakening_proof (new_head_wk, f1, new_tail_wk, new_proof)
```

Comment:
Is this subsumed by the next case?

Reply:
The next case is just swapping weakenings that are above With_proof. This case takes one weakening from both premises and put it below With_proof.